commands-v2/ramsete/subsystems/drivetrain.py

```diff
@@ -0,0 +1,159 @@
+        )
+
+    def getPose(self):
+        """Returns the current position of the robot using it's odometry."""
+        return self.odometry.getPose()
+
+    def getWheelSpeeds(self):
+        """Return an object which represents the wheel speeds of our drivetrain."""
+        speeds = DifferentialDriveWheelSpeeds(
+            self.leftEncoder.getRate(), self.rightEncoder.getRate()
+        )
+        return speeds
+
+    def resetOdometry(self, pose):
+        """ Resets the robot's odometry to a given position."""
+        self.resetEncoders()
+        self.odometry.resetPosition(pose, self.gyro.getRotation2d())
+
+    def arcadeDrive(self):
+        """Drive the robot with standard arcade controls."""
+        self.drive.arcadeDrive(
+            -self.controller.getRawAxis(1),  # Invert the y-axis's input.
+            self.controller.getRawAxis(2) * 0.65,  # Multiply by 65% for more control.
+        )
+
+    def tankDriveVolts(self, leftVolts, rightVolts):
+        """Control the robot's drivetrain with voltage inputs for each side."""
+        self.leftMotors.setVoltage(leftVolts)  # Set the voltage of the left side.
+        self.rightMotors.setVoltage(
+            -rightVolts
+        )  # Set the voltage of the right side. It's
+        # inverted with a negative sign because it's motors need to spin in the negative direction
+        # to move forward.
+
+        self.drive.feed()  # Resets the timer for this motor's MotorSafety
+
+    def stopMoving(self):
+        """Stops the robot from moving. This is important because we need a reference to not take any arguments."""
+        self.leftMotors.setVoltage(0)
+        self.rightMotors.setVoltage(0)
+
+    def resetEncoders(self):
+        """Resets the encoders of the drivetrain."""
+        self.leftEncoder.reset()
+        self.rightEncoder.reset()
+
+    def getAverageEncoderDistance(self):
+        """
+        Take the sum of each encoder's traversed distance and divide it by two,
+        since we have two encoder values, to find the average value of the two.
+        """
+        return (self.leftEncoder.getDistance() + self.rightEncoder.getDistance()) / 2
+
+    def getLeftEncoder(self):
+        """Returns the left encoder object."""
+        return self.leftEncoder
+
+    def getRightEncoder(self):
+        """Returns the right encoder object."""
+        return self.rightEncoder
+
+    def setMaxOutput(self, maxOutput):
+        """Set the max percent output of the drivetrain, allowing for slower control."""
+        self.drive.setMaxOutput(maxOutput)
+
+    def setSlowMaxOutput(self):
+        """Sets the slow max output. Needed a method that doesn't take any arguments."""
+        self.setMaxOutput(0.5)
+
+    def setNormalMaxOutput(self):
+        """Sets the standard max output. Needed a method that doesn't take any arguments."""
+        self.setMaxOutput(1)
+
+    def zeroHeading(self):
+        """Zeroes the gyro's heading."""
+        self.gyro.reset()
```

Comment:
Hmm, this looks like it'd break the odometry tracking. I presume this is an upstream bug though...

Reply:
Ok. Is there anything you want me to do about it? I'm actually not sure if we call it anywhere ... will check later.